matlab/CMakeLists.txt

```diff
@@ -212,56 +233,150 @@ if(MATLAB_BUILD_TESTS)
-set(featherread_sources featherreadmex.cc feather_reader.cc util/handle_status.cc
-                        util/unicode_conversion.cc)
-list(TRANSFORM featherread_sources PREPEND ${CMAKE_SOURCE_DIR}/src/)
-
-# Build featherreadmex MEX binary
-matlab_add_mex(R2018a
-               NAME featherreadmex
-               SRC ${featherread_sources}
-               LINK_TO arrow_shared)
+message(STATUS "Mex Library: ${Matlab_MEX_LIBRARY}")
+message(STATUS "Mex Include Folder: ${Matlab_INCLUDE_DIRS}")
+
+set(CPP_SOURCE_DIR ${CMAKE_SOURCE_DIR}/src/cpp)
+set(MATLAB_SOURCE_DIR ${CMAKE_SOURCE_DIR}/src/cpp/arrow/matlab)
+
+set(arrow_matlab_sources
+    mex/mex_util.cc
+    feather/feather_reader.cc
+    feather/feather_writer.cc
+    feather/feather_functions.cc
+    feather/util/handle_status.cc
+    feather/util/unicode_conversion.cc)
+list(TRANSFORM arrow_matlab_sources PREPEND ${CPP_SOURCE_DIR}/arrow/matlab/)
+
+add_library(arrow_matlab SHARED ${arrow_matlab_sources})
+
+# Declare a dependency on arrow_shared (libarrow.so/dylib/dll).
+target_link_libraries(arrow_matlab arrow_shared)
+
+# On macOS, rpaths for linking imported targets are not automatically added to linked libraries
+if(APPLE)
+  get_property(ARROW_SHARED_IMPORTED_LOCATION
+               TARGET arrow_shared
+               PROPERTY IMPORTED_LOCATION)
+  get_filename_component(ARROW_SHARED_LIBRARY_DIR ${ARROW_SHARED_IMPORTED_LOCATION}
+                         DIRECTORY)
+  target_link_options(arrow_matlab PUBLIC -Wl,-rpath,${ARROW_SHARED_LIBRARY_DIR})
+endif()
 
-# Construct the absolute path to featherwrite's source files
-set(featherwrite_sources featherwritemex.cc feather_writer.cc util/handle_status.cc
-                         util/unicode_conversion.cc)
-list(TRANSFORM featherwrite_sources PREPEND ${CMAKE_SOURCE_DIR}/src/)
+# Declare a dependency on the MEX shared library (libmex.so/dylib/dll).
+target_link_libraries(arrow_matlab ${Matlab_MX_LIBRARY})
+target_link_libraries(arrow_matlab ${Matlab_MEX_LIBRARY})
+
+# Include the MATLAB MEX headers.
+target_include_directories(arrow_matlab PRIVATE ${Matlab_INCLUDE_DIRS})
+target_include_directories(arrow_matlab PRIVATE ${CPP_SOURCE_DIR})
+target_include_directories(arrow_matlab PRIVATE ${ARROW_INCLUDE_DIR})
+target_compile_definitions(arrow_matlab PRIVATE ARROW_MATLAB_EXPORTING)
+
+set(mexfcn_sources mex/mexfcn.cc)
+list(TRANSFORM mexfcn_sources PREPEND ${CPP_SOURCE_DIR}/arrow/matlab/)
+
+# Store the CMake Configuration Type for multi-configuration generators so
+# that we can reference it later to automatically add the correct build
+# folder to the MATLAB Search Path
+get_property(is_multi_config GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
+if(is_multi_config)
+  add_custom_command(TARGET arrow_matlab
+                     POST_BUILD
+                     COMMAND "${CMAKE_COMMAND}" -E echo $<CONFIG> >
+                             ${CMAKE_BINARY_DIR}/is_multi_config.txt)
+endif()
```

Comment:
Can we deffer this to a follow-up pull request to reduce changes in this pull request?

Reply:
This makes sense to me! I have removed the code to create `is_multi_config.txt`. And I will add it in a future pull request for [ARROW-15182](https://issues.apache.org/jira/browse/ARROW-15182).